testsuite/tests/frame-pointers/fp_backtrace.c

```diff
@@ -1,110 +1,156 @@
 {
   struct frame_info*  prev;     /* rbp */
   void*               retaddr;  /* rip */
 } frame_info;
 
-jmp_buf resume_buf;
 
+/*
+ * A backtrace symbol looks like:
+ * ./path/to/binary(camlModule_fn_123+0xAABBCC) [0xAABBCCDDEE]
+ */
+static const char* backtrace_symbol(const struct frame_info* fi)
+{
+  char** symbols = backtrace_symbols(&fi->retaddr, 1);
+  if (!symbols) {
+    perror("backtrace_symbols");
+    return NULL;
+  }
+
+  const char* symbol = strdup(symbols[0]);
+  free(symbols);
+  return symbol;
+}
 
-static void signal_handler(int signum)
+static bool is_from_executable(const char* symbol, const char* execname)
 {
-  /* Should be safe to be called from a signal handler.
-   * See 21.2.1 "Performing a nonlocal goto from a signal handler" from
-   * The Linux Programming Interface, Michael Kerrisk */
-  siglongjmp(resume_buf, 1);
+  return strncmp(symbol, execname, strlen(execname)) == 0;
 }
 
-static int install_signal_handlers(const int signals[], struct sigaction
-    handlers[], int count)
+static regmatch_t func_name_from_symbol(const char* symbol)
 {
-  for (int i = 0; i < count; i++) {
-    struct sigaction action = { 0 };
-    action.sa_handler = signal_handler;
-    sigemptyset(&action.sa_mask);
-    action.sa_flags = 0;
-
-    if (sigaction(signals[i], &action, &handlers[i]) != 0) {
-      perror("sigaction");
-      return -1;
-    }
+  regex_t     regex;
+  regmatch_t  match[2] = { {-1, -1}, {-1, -1}};
+  char        errbuf[128];
+  int         err;
+
+  err = regcomp(&regex, RE_FUNC_NAME, REG_EXTENDED);
+  if (err) {
+    regerror(err, &regex, errbuf, ARR_SIZE(errbuf));
+    fprintf(stderr, "regcomp: %s\n", errbuf);
+    return match[0];
   }
-  return 0;
+
+  err = regexec(&regex, symbol, ARR_SIZE(match), match, 0);
+  if (err == REG_NOMATCH)
+    return match[0];
+
+  return match[1];
 }
 
-static int restore_signal_handlers(const int signals[], struct sigaction
-    handlers[], int count)
+static bool is_caml_entry(const char* symbol, const regmatch_t* funcname)
 {
-  for (int i = 0; i < count; i++) {
-    if (sigaction(signals[i], &handlers[i], NULL) != 0) {
-      perror("sigaction");
-      return -1;
-    }
-  }
-  return 0;
+  //regoff_t len = funcname->rm_eo - funcname->rm_so;
+  //return strnstr(symbol + funcname->rm_so, CAML_ENTRY, len) == 0;
+  return strstr(symbol + funcname->rm_so, CAML_ENTRY) != NULL;
 }
 
-static int safe_read(const struct frame_info* fi, struct frame_info** prev,
-    void** retaddr)
+static regmatch_t trim_func_name(const char* symbol, const regmatch_t* funcname)
 {
-  /* Signals to ignore while attempting to read frame_info members */
-  const int signals[] = { SIGSEGV, SIGBUS };
-  /* Store original signal handers */
-  struct sigaction handlers[ARRSIZE(signals)] = { 0 };
-  int ret = 0;
-
-  if (install_signal_handlers(signals, handlers, ARRSIZE(signals)) != 0)
-    return -1;
-
-  if (!sigsetjmp(resume_buf, 1)) {
-    *prev = fi->prev;
-    *retaddr = fi->retaddr;
```

Comment:
Do I read correctly, that the signal mask and long jump business in `safe_read` was here to handle the read to `retaddr`, and this is why the present PR remove the function (and it auxiliary functions)?

Reply:
Yes, this is absolutely correct.
Previously, `fp_backtrace` tried to unwind the whole stack, but the linked list of 'saved rbp' was not always `NULL` terminated, as I expected. `safe_read` installed a temporary signal handler to gracefully handle an invalid pointer deference attempt.

Stopping the backtrace (and thus stack unwinding) at `caml_program` means this dubious trick is no longer needed.